src/gmcp.cpp

```diff
@@ -0,0 +1,404 @@
+/******************************************************************************
+ Header files.
+ ******************************************************************************/
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <arpa/telnet.h>
+#include <time.h>
+#include <ctype.h>
+
+#include "gmcp.hpp"
+#include "utils.hpp"
+#include "awake.hpp"
+#include "structs.hpp"
+#include "constants.hpp"
+#include "comm.hpp"
+#include "protocol.hpp"
+#include "config.hpp"
+#include "nlohmann/json.hpp"
+#include "db.hpp"
+#include "protocol.hpp"
+#include "newmatrix.hpp"
+
+using nlohmann::json;
+
+static void Write( descriptor_t *apDescriptor, const char *apData )
+{
+  if ( apDescriptor != NULL ) {
+   write_to_descriptor( apDescriptor->descriptor, apData );
+  }
+}
+
+/******************************************************************************
+ GMCP functions.
+ ******************************************************************************/
+void SendGMCPCoreSupports ( descriptor_t *apDescriptor )
+{
+  if (!apDescriptor || !apDescriptor->pProtocol->bGMCP) return;
+  json j;
+
+  j["Core"] = json::array();
+  j["Core"].push_back("Supports");
+
+  j["Room"] = json::array();
+  j["Room"].push_back("Info");
+  j["Room"].push_back("Exits");
+
+  j["Char"] = json::array();
+  j["Char"].push_back("Vitals");
+  j["Char"].push_back("Info");
+  j["Char"].push_back("Pools");
+
+  j["Matrix"] = json::array();
+  j["Matrix"].push_back("Info");
+  j["Matrix"].push_back("Deck");
+
+  // Dump the json to a string and send it.
+  std::string payload = j.dump();
+  SendGMCP(apDescriptor, "Core.Supports", payload.c_str());
+}
+
+void SendGMCPMatrixInfo ( struct char_data *ch )
+{
+  if (!ch->desc || !ch->desc->pProtocol->bGMCP) return;
+  struct matrix_icon *persona = ch->persona;
+
+  json j;
+
+  if (!persona || !persona->in_host || !persona->decker) {
+    std::string payload = j.dump();
+    SendGMCP(ch->desc, "Matrix.Info", payload.c_str());
+    return;
+  }
+
+  struct host_data *cur_host = &matrix[persona->in_host];
+  j["name"] = cur_host->name;
+  j["description"] = cur_host->desc;
+  j["vnum"] = cur_host->vnum;
+  j["color"] = cur_host->color;
```

Comment:
I think you usually have to run a program to see this, which gives a risk of being spotted by IC.